sklearn/ensemble/_iforest.py

```diff
@@ -509,6 +522,25 @@ def _more_tags(self):
         }
 
 
+def _compute_depths(tree):
```

Comment:
I think that this function belong to the tree code. I would implement it in cython:

```python
def compute_depths(object tree):
    cdef:
        cnp.int64_t[::1] depths = np.empty(tree.node_count, dtype=np.int64)
        cnp.npy_intp[:] children_left = tree.children_left
        cnp.npy_intp[:] children_right = tree.children_right
        cnp.npy_intp TREE_LEAF = -1
        cnp.npy_intp node_count = tree.node_count
        cnp.npy_intp node_id
        cnp.int64_t depth
        
    # init root node
    depths[0] = 1
    for node_id in range(node_count):
        if children_left[node_id] != TREE_LEAF:
            depth = depths[node_id] + 1
            depths[children_left[node_id]] = depth
            depths[children_right[node_id]] = depth
    
    return depths.base
```

I did a quick benchmark on the following tree:

```python
from sklearn.datasets import make_classification
from sklearn.tree import DecisionTreeClassifier

X, y = make_classification(n_samples=10_000, n_features=10)
clf = DecisionTreeClassifier(random_state=0).fit(X, y)
```

We go from

```
1.07 ms ± 27.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

to

```
3.23 µs ± 188 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

and the code is really similar so I think this is worth doing it.

Reply:
Note that you need to use the `TREE_LEAF` already available in the `_tree.pyx`.

Reply:
Moved to Cython